### PR TITLE
Move to a unified SyntaxList model for C# and VB.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler  for review.

This PR has been perf tested and has revealed no issues.   Overall there are ~1650 lines of essentially duplicated code that we are able to remove because of this.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler .  This is now ready for review.  I don't have a great recommendation on how to get through this.  You could go commit by commit to keep each change small.  or you can try to get through it in one fell swoop.

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs [hunk 1]
```diff
@@ -36,7 +37,7 @@ private DeclarationTreeBuilder(SyntaxTree syntaxTree, string scriptClassName, bo
         private ImmutableArray<SingleNamespaceOrTypeDeclaration> VisitNamespaceChildren(
             CSharpSyntaxNode node,
             SyntaxList<MemberDeclarationSyntax> members,
-            Syntax.InternalSyntax.SyntaxList<Syntax.InternalSyntax.MemberDeclarationSyntax> internalMembers)
+            CommonSyntaxList<Syntax.InternalSyntax.MemberDeclarationSyntax> internalMembers)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs]: Can `CommonSyntaxList` be renamed to `SyntaxList`?

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs [hunk 2]
```diff
@@ -526,7 +528,7 @@ private static void AddNonTypeMemberNames(Syntax.InternalSyntax.CSharpSyntaxNode
             {
                 case SyntaxKind.FieldDeclaration:
                     anyNonTypeMembers = true;
-                    Syntax.InternalSyntax.SeparatedSyntaxList<Syntax.InternalSyntax.VariableDeclaratorSyntax> fieldDeclarators =
+                    CommonSeparatedSyntaxList<Syntax.InternalSyntax.VariableDeclaratorSyntax> fieldDeclarators =
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs]: Rename `CommonSeparatedSyntaxList` to `SeparatedSyntaxList`?

File: src/Compilers/CSharp/Portable/Parser/SyntaxFactoryContext.cs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/SyntaxFactoryContext.cs]: Revert.

File: src/Compilers/CSharp/Portable/PublicAPI.Shipped.txt [hunk 4]
```diff
@@ -3266,8 +3263,6 @@ override Microsoft.CodeAnalysis.CSharp.Syntax.XmlTextSyntax.Accept(Microsoft.Cod
 override Microsoft.CodeAnalysis.CSharp.Syntax.XmlTextSyntax.Accept<TResult>(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor<TResult> visitor) -> TResult
 override Microsoft.CodeAnalysis.CSharp.Syntax.YieldStatementSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor visitor) -> void
 override Microsoft.CodeAnalysis.CSharp.Syntax.YieldStatementSyntax.Accept<TResult>(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor<TResult> visitor) -> TResult
-override sealed Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode.ToFullString() -> string
-override sealed Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode.ToString() -> string
```
Comment (cston) [inline src/Compilers/CSharp/Portable/PublicAPI.Shipped.txt]: Are these breaking changes?

File: src/Compilers/CSharp/Portable/Syntax/InternalSyntax/SyntaxNode.cs [hunk 5]
```diff
@@ -184,43 +146,49 @@ public override GreenNode GetTrailingTriviaCore()
...
                     }
                 }
             }
 
             return stack;
         }
 
+        internal static DirectiveStack ApplyDirectivesToListOrNode(GreenNode listOrNode, DirectiveStack stack)
+        {
+            // If we have a list of trivia, then that node is not actually a CSharpSyntaxNode.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/InternalSyntax/SyntaxNode.cs]: Some internal nodes are no longer `CSharpSyntaxNode`s? Are all callers expecting that?

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/InternalSyntax/SyntaxNode.cs]: To the best of my abilities, all callers are expecting the right information.

File: src/Compilers/CSharp/Portable/PublicAPI.Shipped.txt (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/PublicAPI.Shipped.txt]: No.  Some of the overrides were removed because the core functionality was moved up to the base class.  The calls are still virtual, it's just not necessayr for VB and C# to implement htem in the exact same fashion.

File: src/Compilers/CSharp/Portable/Syntax/InternalSyntax/SyntaxToken.cs [hunk 6]
```diff
@@ -379,17 +374,18 @@ internal override DirectiveStack ApplyDirectives(DirectiveStack stack)
...
 
-                var trailing = this.GetTrailingTrivia();
-                if (trailing != null && trailing.ContainsDirectives)
-                {
-                    stack = trailing.ApplyDirectives(stack);
-                }
+            return stack;
+        }
+
+        private DirectiveStack ApplyDirectivesToTrivia(GreenNode triviaList, DirectiveStack stack)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/InternalSyntax/SyntaxToken.cs]: `static`?

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 7]
```diff
@@ -566,36 +621,116 @@ internal DiagnosticInfo[] GetDiagnostics()
...
+                }
+
+                if (currentNode.IsTrivia)
+                {
+                    currentNode.WriteTriviaTo(writer);
+                    continue;
+                }
+
+                var firstIndex = GetFirstNonNullChildIndex(currentNode);
+                var lastIndex = GetLastNonNullChildIndex(currentNode);
```
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs]: It looks like `GetFirst/LastNonNullChildIndex` are similar to the loop below. Is this the only place we're using those methods? If so, consider removing those methods and looping over all items here instead.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 8]
```diff
@@ -566,36 +621,116 @@ internal DiagnosticInfo[] GetDiagnostics()
...
+
+        protected virtual void WriteTriviaTo(TextWriter writer)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected virtual void WriteTokenTo(TextWriter writer, bool leading, bool trailing)
+        {
+            throw new NotImplementedException();
         }
```
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs]: Could these be `abstract` methods instead?

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 9]
```diff
@@ -829,5 +993,35 @@ internal bool IsCacheEquivalent(int kind, NodeFlags flags, GreenNode child1, Gre
...
+
+            // If the green node already has errors, add those on.
+            if (GetDiagnostics() == null)
+            {
+                errorInfos = new[] { err };
+            }
+            else
+            {
+                // Add the error to the error list.
+                errorInfos = GetDiagnostics();
```
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs]: Consider moving ahead of `if`:
```
var errorInfos = GetDiagnostics();
if (errorInfos == null)
{
    ...
}
else
{
    ...
}
```

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs [hunk 10]
```diff
@@ -59,13 +57,18 @@ public void Clear()
             _builder.Clear();
         }
 
+        public void RemoveLast()
+        {
+            this._builder.RemoveLast();
+        }
```
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs]: Is this method used?

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs [hunk 11]
```diff
@@ -1,11 +1,13 @@
...
 
 using System;
+using System.Collections;
+using System.Collections.Generic;
 
 namespace Microsoft.CodeAnalysis.Syntax.InternalSyntax
 {
     //This used to implement IEnumerable and have more functionality, but it was not tested or used.
-    internal struct SyntaxDiagnosticInfoList
+    internal struct SyntaxDiagnosticInfoList : IEnumerable<DiagnosticInfo>
```
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: Is the `IEnumerable<DiagnosticInfo>` implementation used? If not, remove.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs (hunk 8 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs]: I avoided abstract here to make it so we wouldn't have a bunch of methods to implement all over the place.

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs (hunk 11 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: Yup. VB uses that.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs (hunk 9 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs]: I copied the VB version verbatim. Trying not to make unnecessary changes that cause hard to track down bugs :-)

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs (hunk 8 above)
Comment (tmat) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs]: I'd use ```throw ExceptionUtilities.Unreachable``` then.

**Comment by CyrusNajmabadi** [comment]
I can definitely do that. Btw, do you still want me to squash?  I'm happy to.  However, since you started reviewing, i wasn't sure.

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs]: Yes. By VB.  These files are the merged forms of the c# and VB impls

Comment (cston) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs]: It doesn't appear to be used by VB.

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs (hunk 11 above)
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: It doesn't appear to be used by VB.

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs]: This is used here: 
https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Syntax/InternalSyntax/SyntaxRewriter.vb#L88

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs (hunk 11 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: This is used here (and other places):
https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Test/Syntax/Scanner/ScannerTests.vb#L703

VB calls linq extension methods (i.e. Count) on the SyntaxDiagnosticInfoList

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs (hunk 10 above)
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SeparatedSyntaxListBuilder.cs]: I see, thanks. (I was looking at `Syntax.SeparatedSyntaxListBuilder` rather than `Syntax.InternalSyntax...`.)

File: src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs (hunk 11 above)
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: The VB version of `SyntaxDiagnosticInfoList` contained a `Count` property that was used from ScannerTests.vb. From the VB version, after commenting out the `IEnumerable(Of DiagnosticInfo)` implementation and the private `Nodes` property, the solution compiles.

Comment (cston) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: That said, the `Count` property previously was complex and appears to be called from ScannerTests.vb only. If that's the case, the `Count` property could be removed and the tests could use the explicit `GetEnumerator` method to count the items.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxDiagnosticInfoList.cs]: You're correct.  I missed that.  Given that this property is only used for test code, i'm just going to make a test helper that does things in a simple manner.  THanks!

**Comment by CyrusNajmabadi** [comment]
Retest microbuild_prtest  please
